pkg/gui/command_status_panel.go

```diff
@@ -0,0 +1,66 @@
+package gui
+
+import (
+	"fmt"
+
+	"github.com/jesseduffield/gocui"
+	"github.com/scbizu/cmd"
+)
+
+// CommandStatus defines the go-cmd's cmd
+type CommandStatus struct {
+	command *cmd.Cmd
+	gui     *gocui.Gui
+}
+
+// NewCommandStatus new command status
+func NewCommandStatus(c *cmd.Cmd, gui *gocui.Gui) *CommandStatus {
+	return &CommandStatus{
+		command: c,
+		gui:     gui,
+	}
+}
+
+// Update updates the command status panel
+func (cs *CommandStatus) Update(ui *Gui) {
+
+	go func() {
+		for {
+			select {
+			case status := <-cs.command.Stdout:
+				ui.Log.Infof("collect message:%v", status)
+				if err := cs.refreshCommandStatus(cs.gui, status); err != nil {
+					ui.Log.Errorf("get commandStatus Panel view failed:%v", err.Error())
+					return
+				}
+			case status := <-cs.command.Stderr:
+				ui.Log.Infof("collect error message:%v", status)
+				if err := cs.refreshCommandStatus(cs.gui, status); err != nil {
+					ui.Log.Errorf("get commandStatus Panel view failed:%v", err.Error())
+					return
+				}
+			}
+		}
+	}()
+
+	<-cs.command.Start()
```

Comment:
I'm not sure what it means to have `<-` here

Reply:
cs.command.Start() return a <-channel, without <- it will block there.

Reply:
reading this: https://github.com/go-cmd/cmd#channel-based-fire-and-forget
I think what's actually happening is that without `<-` it does not block and then an error occurs because two git processes attempt to run simultaneously when the first one has obtained a lock. So having `<-` forces us to wait for the command to be done before we continue.

This isn't ideal if you have pre-commits that take more than a few seconds to run, because you won't be able to watch them stream bit by bit e.g. a test pre-commit file like this:
```
echo 'hello1'
sleep 1
echo 'hello2'
sleep 1
echo 'hello3'
```

I did some playing around and found a way to get streaming to work without having to pause the application. This patch does the trick for me:

```
diff --git a/pkg/gui/command_status_panel.go b/pkg/gui/command_status_panel.go
index 3975c0e..7ce575e 100644
--- a/pkg/gui/command_status_panel.go
+++ b/pkg/gui/command_status_panel.go
@@ -44,7 +44,18 @@ func (cs *CommandStatus) PrintCmdOutput(gui *Gui) {
                }
        }()

-       <-cs.command.Start()
+       statusChan := cs.command.Start()
+
+       go func() {
+               <-statusChan // wait for command to finish
+
+               if err := cs.refreshCommandStatus(cs.gui, "\n"); err != nil {
+                       gui.Log.Errorf("get commandStatus Panel view failed:%v", err.Error())
+               }
+
+               gui.refreshFiles(gui.g)
+               gui.refreshCommits(gui.g)
+       }()
 }

 func (cs *CommandStatus) refreshCommandStatus(g *gocui.Gui, cmdStr string) error {
diff --git a/pkg/gui/commit_message_panel.go b/pkg/gui/commit_message_panel.go
index b8f559f..2325600 100644
--- a/pkg/gui/commit_message_panel.go
+++ b/pkg/gui/commit_message_panel.go
@@ -13,6 +13,11 @@ func (gui *Gui) handleCommitConfirm(g *gocui.Gui, v *gocui.View) error {
                return gui.createErrorPanel(g, gui.Tr.SLocalize("CommitWithoutMessageErr"))
        }

+       v.Clear()
+       v.SetCursor(0, 0)
+       g.SetViewOnBottom("commitMessage")
+       gui.switchFocus(g, v, gui.getFilesView(g))
+
        c, isGPG := gui.GitCommand.CommitWithStatus(message)
        if isGPG {
                // put it into subprocess
@@ -30,13 +35,10 @@ func (gui *Gui) handleCommitConfirm(g *gocui.Gui, v *gocui.View) error {
                if c != nil {
                        status := NewCommandStatus(c, g)
                        status.PrintCmdOutput(gui)
+                       return nil
                }
        }
        gui.refreshFiles(g)
-       v.Clear()
-       v.SetCursor(0, 0)
-       g.SetViewOnBottom("commitMessage")
-       gui.switchFocus(g, v, gui.getFilesView(g))
        return gui.refreshCommits(g)
 }
```

Reply:
Though I now realise that that patch messes with things when we have `commandStatus: false`. In the case of people with that value set to false, we don't want the commit message panel to disappear until after the commit is complete, otherwise it's hard to know what's going on. So we don't want to hide the commit message panel prematurely. Hmm

Reply:
This implement is really a trick. I reproduce this bug after I  add `sleep` in my pre-commit.And this is not the correct behavior I expected . With the cmd `Streaming:true` mode, the `<-cs.command.Stdout` should work as **whenever a message in, it will receive this message**. The [example](https://github.com/go-cmd/cmd/blob/master/examples/blocking-streaming/main.go) in cmd-go does this perfectly. Anyway, I will continue to fix this. 

For your another advise, I think It's OK to keep the focus on the streaming status panel when the streaming panel is running.

Reply:
I found it may be blocked at `refreshCommandStatus` when I debug the streaming status.Here are some debug logs:
```
time="2018-10-18T11:26:14+08:00" level=info msg="collect error message:Pre-commit Test Flag #0" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:15+08:00" level=info msg="collect error message:Pre-commit Test Flag #1" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:16+08:00" level=info msg="collect error message:Pre-commit Test Flag #2" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:17+08:00" level=info msg="collect error message:Pre-commit Test Flag #3" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:18+08:00" level=info msg="collect error message:Pre-commit Test Flag #4" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:19+08:00" level=info msg="collect error message:Pre-commit Test Flag #5" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:20+08:00" level=info msg="collect error message:Pre-commit Test Flag #6" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:21+08:00" level=info msg="collect error message:Pre-commit Test Flag #7" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:22+08:00" level=info msg="collect error message:Pre-commit Test Flag #8" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:23+08:00" level=info msg="collect error message:Pre-commit Test Flag #9" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:24+08:00" level=info msg="collect error message:Pre-commit Test Flag #10" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:25+08:00" level=info msg="collect error message:Pre-commit Test Flag #11" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:26+08:00" level=info msg="collect error message:Pre-commit Test Flag #12" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:27+08:00" level=info msg="collect error message:Pre-commit Test Flag #13" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:28+08:00" level=info msg="collect error message:Pre-commit Test Flag #14" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:29+08:00" level=info msg="collect message:[feature/go-cmd 8155b41] Rollback" buildDate= commit= debug=true version=unversioned
time="2018-10-18T11:26:29+08:00" level=info msg="collect message: 1 file changed, 9 insertions(+), 20 deletions(-)" buildDate= commit= debug=true version=unversioned
```